analytics/lib/data_collectors.py

```diff
@@ -0,0 +1,266 @@
+        elif count_type == 'cross_table_aggregate':
+            action = do_aggregate_to_summary_table
+        else:
+            raise ValueError('Unknown count_type')
+        action(stat, time_interval, **kwargs)
+
+# only two summary tables at the moment: RealmCount and DeploymentCount.
+# will have to generalize this a bit if more are added
+
+
+def do_aggregate_to_summary_table(stat, time_interval, from_table, to_table):
+    if to_table == RealmCount:
+        id_cols = 'realm_id,'
+        group_by = 'GROUP BY realm_id'
+    elif to_table == DeploymentCount:
+        id_cols = ''
+        group_by = ''
+    else:
+        raise ValueError("%s is not a summary table" % (to_table,))
+
+    query = """
+        INSERT INTO %(to_table)s (%(id_cols)s value, property, end_time, interval)
+        SELECT %(id_cols)s COALESCE (sum(value), 0), '%(property)s', %%s, '%(interval)s'
+        FROM %(from_table)s WHERE
+        (
+            property = '%(property)s' AND
+            end_time = %%s AND
+            interval = '%(interval)s'
+        )
+        %(group_by)s
+    """ % {'to_table': to_table._meta.db_table,
+           'id_cols' : id_cols,
+           'from_table' : from_table._meta.db_table,
+           'property' : stat.property,
+           'interval' : time_interval.interval,
+           'group_by' : group_by}
+    cursor = connection.cursor()
+    # todo: check if cursor supports the %(key)s syntax
+    cursor.execute(query, (time_interval.start, time_interval.end))
+    cursor.close()
+
+def do_aggregate_hour_to_day(stat, end_time):
+    table = stat.analytics_table
+    id_cols = ''.join([col + ', ' for col in table.extended_id()])
+    group_by = 'GROUP BY %s' % id_cols if id_cols else ''
+    query = """
+        INSERT INTO %(table)s (%(id_cols)s value, property, end_time, interval)
+        SELECT %(id_cols)s sum(value), '%(property)s', %%s, 'day'
+        FROM %(table)s WHERE
+        (
+            property = '%(property)s' AND
+            end_time > %%s AND
+            end_time <= %%s AND
+            interval = 'hour'
+        )
+        %(group_by)s property
+    """ % {'table': table._meta.db_table,
+           'id_cols' : id_cols,
+           'group_by' : group_by,
+           'property': stat.property}
+    cursor = connection.cursor()
+    cursor.execute(query, (end_time.end, end_time.end - timedelta(days=1), end_time.end))
+    cursor.close()
+
+## methods that hit the prod databases directly
+# No left joins in Django ORM yet, so have to use raw SQL :(
+# written in slightly more than needed generality, to reduce copy-paste errors
+# as more of these are made / make it easy to extend to a pull_X_by_realm
+
+def do_pull_from_zerver(stat, time_interval, query):
+    zerver_table = stat.zerver_table._meta.db_table
+    join_args = ' '.join('AND %s.%s = %s' % (zerver_table, key, value) \
+                         for key, value in stat.filter_args.items())
+    query_ = query % {'zerver_table' : zerver_table,
```

Comment:
We should probably add a comment saying why we do the string replacement in two places (here and in cursor.execute). (It's because the join_args can't be done in cursor.execute, and for the time stuff it's convenient to let cursor put it into the required database format, I can explain more when I'm in.)